lua/telescope/builtin/git.lua

```diff
@@ -0,0 +1,177 @@
+local utils = require('telescope.utils')
+
+local conf = require('telescope.config').values
+
+local git = {}
+
+git.files = function(opts)
+  local show_untracked = utils.get_default(opts.show_untracked, true)
+
+  -- By creating the entry maker after the cwd options,
+  -- we ensure the maker uses the cwd options when being created.
+  opts.entry_maker = opts.entry_maker or make_entry.gen_from_file(opts)
+
+  pickers.new(opts, {
+    prompt_title = 'Git File',
+    finder = finders.new_oneshot_job(
+      { "git", "ls-files", "--exclude-standard", "--cached", show_untracked and "--others" },
+      opts
+    ),
+    previewer = previewers.cat.new(opts),
+    sorter = conf.file_sorter(opts),
+  }):find()
+end
+
+git.commits = function(opts)
+  local cmd = 'git log --pretty=oneline --abbrev-commit'
+  local results = vim.split(utils.get_os_command_output(cmd), '\n')
+
+  pickers.new(opts, {
+    prompt_title = 'Git Commits',
+    finder = finders.new_table {
+      results = results,
+      entry_maker = make_entry.gen_from_git_commits(opts),
+    },
+    previewer = previewers.git_commit_diff.new(opts),
+    sorter = conf.file_sorter(opts),
+    attach_mappings = function()
+      actions.goto_file_selection_edit:replace(actions.git_checkout)
+      return true
+    end
+  }):find()
+end
+
+git.bcommits = function(opts)
+  local cmd = 'git log --pretty=oneline --abbrev-commit ' .. vim.fn.expand('%')
+  local results = vim.split(utils.get_os_command_output(cmd), '\n')
+
+  pickers.new(opts, {
+    prompt_title = 'Git BCommits',
+    finder = finders.new_table {
+      results = results,
+      entry_maker = make_entry.gen_from_git_commits(opts),
+    },
+    previewer = previewers.git_commit_diff.new(opts),
+    sorter = conf.file_sorter(opts),
+    attach_mappings = function()
+      actions.goto_file_selection_edit:replace(actions.git_checkout)
+      return true
+    end
+  }):find()
+end
+
+git.branches = function(opts)
+  -- Does this command in lua (hopefully):
+  -- 'git branch --all | grep -v HEAD | sed "s/.* //;s#remotes/[^/]*/##" | sort -u'
+  local output = vim.split(utils.get_os_command_output('git branch --all'), '\n')
+
+  local tmp_results = {}
+  for _, v in ipairs(output) do
+    if not string.match(v, 'HEAD') and v ~= '' then
+      v = string.gsub(v, '.* ', '')
+      v = string.gsub(v, '^remotes/.*/', '')
+      tmp_results[v] = 1
+    end
+  end
+
+  local results = {}
+  for k, _ in pairs(tmp_results) do
+    table.insert(results, k)
+  end
+
+  pickers.new(opts, {
+    prompt_title = 'Git Branches',
+    finder = finders.new_table {
+      results = results,
+      entry_maker = function(entry)
+        return {
+          value = entry,
+          ordinal = entry,
+          display = entry,
+        }
+      end
+    },
+    previewer = previewers.git_branch_log.new(opts),
+    sorter = conf.file_sorter(opts),
+    attach_mappings = function()
+      actions.goto_file_selection_edit:replace(actions.git_checkout)
```

Comment:
it's so pretty :)

Reply:
I agree, so simple and pretty. Thanks for that great new api